File: dart-quic-ffi/src/binary_protocol.rs
/// 
/// High-efficiency binary protocol - for Rust and Dart FFI communication
/// 

/// Protocol version - for backward compatibility
pub const PROTOCOL_VERSION: u8 = 1;

/// Protocol magic number
pub const PROTOCOL_MAGIC: u32 = 0xDABCFE01;

/// Task status
#[repr(u16)]
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TaskStatus {
    // === 0x0000 - 0x00FF ===
    Success = 0x0000,          // Task completed successfully, no return data
    SuccessWithData = 0x0001,  // Task completed successfully, has return data
    
    // === 0x0100 - 0x01FF ===
    WorkerShutdown = 0x0100,   // Worker thread shutdown normally

    
    // === 0x9000 - 0x9FFF ===
    UnknownError = 0x9001,     // Unknown error
    
    // === 0xF000 - 0xFFFF ===
    ProtocolError = 0xF001,    // Protocol error
    VersionMismatch = 0xF002,  // Version mismatch
    CorruptedData = 0xF003,    // Data corrupted
}

/// Data type - represents the specific type of data in the payload
#[repr(u8)]
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum DataType {
    None = 0,        // No data
    Bool = 1,        // Boolean value
    U64 = 2,         // 64-bit unsigned integer
    Bytes = 3,       // Byte array (zero-copy pointer)
    String = 4,      // String (error messages, etc.)
}

/// High-efficiency message header - 16 bytes fixed size, cache-friendly
#[repr(C)]
#[derive(Debug, Clone, Copy)]
pub struct MessageHeader {
    /// Protocol magic number (4 bytes) 
    pub magic: u32,
    /// Protocol version (1 byte)
    pub version: u8,
    /// Data type (1 byte)
    pub data_type: DataType,
    /// Task status (2 bytes) 
    pub status: TaskStatus,
    /// Task ID (8 bytes)
    pub task_id: u64,
}

/// Data payload - corresponds to data types in CommandResult
#[repr(C)]
pub union DataPayload {
    // Basic types - directly embedded, zero-copy
    pub bool_val: bool,
    pub u64_val: u64,
    
    // Complex types - pointer + length
    pub bytes: BytesData,   // Corresponds to CommandResult::WithData
    pub string: StringData, // Corresponds to CommandResult::Error
}

/// Byte data structure
#[repr(C)]
#[derive(Debug, Clone, Copy)]
pub struct BytesData {
    pub ptr: *mut u8,
    pub len: usize,
}

/// String data structure (UTF-8)
#[repr(C)]
#[derive(Debug, Clone, Copy)]
pub struct StringData {
    pub ptr: *const u8,
    pub len: usize,
}

/// Complete task event message structure - 16-byte header + 16-byte payload = 32 bytes
#[repr(C)]
pub struct TaskEventMessage {
    /// Message header (16 bytes)
    pub header: MessageHeader,
    /// Data payload (16-byte union)
    pub payload: DataPayload,
}

impl TaskEventMessage {
    /// Create no-data success message (corresponds to CommandResult::NoData)
    pub fn no_data(task_id: u64) -> Self {
        Self {
            header: MessageHeader {
                magic: PROTOCOL_MAGIC,
                version: PROTOCOL_VERSION,
                status: TaskStatus::Success,
                data_type: DataType::None,
                task_id,
            },
            payload: DataPayload { u64_val: 0 },
        }
    }

    /// Create boolean value message (corresponds to CommandResult::Bool)
    pub fn bool_data(task_id: u64, value: bool) -> Self {
        Self {
            header: MessageHeader {
                magic: PROTOCOL_MAGIC,
                version: PROTOCOL_VERSION,
                status: TaskStatus::SuccessWithData,
                data_type: DataType::Bool,
                task_id,
            },
            payload: DataPayload { bool_val: value },
        }
    }

    /// Create U64 message (corresponds to CommandResult::U64)
    pub fn u64_data(task_id: u64, value: u64) -> Self {
        Self {
            header: MessageHeader {
                magic: PROTOCOL_MAGIC,
                version: PROTOCOL_VERSION,
                status: TaskStatus::SuccessWithData,
                data_type: DataType::U64,
                task_id,
            },
            payload: DataPayload { u64_val: value },
        }
    }

    /// Create byte data message (corresponds to CommandResult::WithData)
    pub fn bytes_data(task_id: u64, data_ptr: *mut u8, data_len: usize) -> Self {
        Self {
            header: MessageHeader {
                magic: PROTOCOL_MAGIC,
                version: PROTOCOL_VERSION,
                status: TaskStatus::SuccessWithData,
                data_type: DataType::Bytes,
                task_id,
            },
            payload: DataPayload {
                bytes: BytesData {
                    ptr: data_ptr,
                    len: data_len,
                },
            },
        }
    }

    /// Create string data message (general string transmission, including error messages)
    pub fn string_data(task_id: u64, status: TaskStatus, text: &str) -> Self {
        let text_bytes = text.as_bytes();
        Self {
            header: MessageHeader {
                magic: PROTOCOL_MAGIC,
                version: PROTOCOL_VERSION,
                status,
                data_type: DataType::String,
                task_id,
            },
            payload: DataPayload {
                string: StringData {
                    ptr: text_bytes.as_ptr(),
                    len: text_bytes.len(),
                },
            },
        }
    }

    /// Convenience method for creating error messages
    pub fn error_message(task_id: u64, error_type: TaskStatus, error_msg: &str) -> Self {
        Self::string_data(task_id, error_type, error_msg)
    }

    /// Create worker thread shutdown message
    pub fn shutdown_message() -> Self {
        Self {
            header: MessageHeader {
                magic: PROTOCOL_MAGIC,
                version: PROTOCOL_VERSION,
                status: TaskStatus::WorkerShutdown,
                data_type: DataType::None,
                task_id: 0,
            },
            payload: DataPayload { u64_val: 0 },
        }
    }

    /// Check if the message is successful
    pub fn is_success(&self) -> bool {
        matches!(self.header.status, TaskStatus::Success | TaskStatus::SuccessWithData)
    }

    /// Check if the message is an error
    pub fn is_error(&self) -> bool {
        !self.is_success() && self.header.status != TaskStatus::WorkerShutdown
    }

    /// Validate message integrity
    pub fn is_valid(&self) -> bool {
        self.header.magic == PROTOCOL_MAGIC && self.header.version == PROTOCOL_VERSION
    }

    /// Get total message size (fixed 32 bytes, zero-copy design)
    pub fn total_size(&self) -> usize {
        // All messages are fixed 32 bytes, data is passed through pointers for zero-copy
        std::mem::size_of::<TaskEventMessage>()
    }
}

/// High-performance serializer - based on TLV design
pub struct MessageSerializer;

impl MessageSerializer {
    /// Serialize message to binary data - fixed 32-byte zero-copy format
    /// 
    /// Format: [Header:16 bytes][Payload:16 bytes] = fixed 32 bytes
    /// 
    /// Design principles:
    /// 1. All messages are fixed 32 bytes for fast transmission and parsing
    /// 2. Basic types (bool, u64) are directly embedded in Payload
    /// 3. Complex types (Bytes, String) are passed through pointers for zero-copy
    /// 4. No copying of long data, maintains high performance
    pub fn serialize(msg: &TaskEventMessage) -> Vec<u8> {
        let msg_size = std::mem::size_of::<TaskEventMessage>();
        let mut buffer = Vec::with_capacity(msg_size);
        
        // Serialize the entire message structure (32 bytes)
        unsafe {
            let msg_bytes = std::slice::from_raw_parts(
                msg as *const _ as *const u8,
                msg_size,
            );
            buffer.extend_from_slice(msg_bytes);
        }
        
        // Note: Do not copy data pointed to by pointers, maintaining zero-copy characteristics
        // DataType::Bytes and DataType::String actual data is referenced by pointers
        // Dart side needs to access original data based on pointer and length
        
        buffer
    }

    /// Deserialize binary data to message
    /// 
    /// # Safety
    /// Caller must ensure data points to valid 32-byte message data
    pub unsafe fn deserialize(data: *const u8, len: usize) -> Option<TaskEventMessage> {
        let msg_size = std::mem::size_of::<TaskEventMessage>();
        if len < msg_size {
            return None;
        }

        // Read fixed 32-byte message
        unsafe {
            let msg = std::ptr::read(data as *const TaskEventMessage);
            
            // Validate protocol integrity
            if !msg.is_valid() {
                return None;
            }
            
            Some(msg)
        }
    }
    
    /// Get data pointed to by pointer (zero-copy access)
    /// 
    /// Note: Data pointed to by pointer is not in the serialization buffer, but at original location
    /// This is the core of zero-copy design - only pass pointers, don't copy data
    /// 
    /// # Safety
    /// Caller must ensure:
    /// 1. Pointer is valid during usage
    /// 2. Memory region pointed to by pointer is readable
    /// 3. Data length is correct
    pub unsafe fn get_data_pointer(msg: &TaskEventMessage) -> Option<(*const u8, usize)> {
        match msg.header.data_type {
            DataType::Bytes => {
                unsafe {
                    let bytes_data = &msg.payload.bytes;
                    if !bytes_data.ptr.is_null() && bytes_data.len > 0 {
                        Some((bytes_data.ptr, bytes_data.len))
                    } else {
                        None
                    }
                }
            }
            DataType::String => {
                unsafe {
                    let string_data = &msg.payload.string;
                    if !string_data.ptr.is_null() && string_data.len > 0 {
                        Some((string_data.ptr, string_data.len))
                    } else {
                        None
                    }
                }
            }
            _ => None,
        }
    }
}

/// Message sender - properly handles Dart memory model
pub struct MessageSender;

impl MessageSender {
    /// Send message to Dart - using Vec<u8>'s IntoDart implementation
    /// 
    pub fn send_to_dart(dart_port: i64, msg: TaskEventMessage) -> bool {
        // Serialize message to byte array
        let binary_data = MessageSerializer::serialize(&msg);
        
        // Send Vec<u8> directly, allo_isolate will automatically handle Dart_CObject creation and memory management
        use allo_isolate::Isolate;
        let isolate = Isolate::new(dart_port);
        isolate.post(binary_data) // Vec<u8> implements IntoDart
    }
}

// Compile-time assertions - ensure struct sizes meet expectations
const _: () = {
    assert!(std::mem::size_of::<MessageHeader>() == 16);
    // DataPayload is a union, size is the size of the largest member (BytesData or StringData = 16 bytes)
    assert!(std::mem::size_of::<DataPayload>() == 16);
    // TaskEventMessage total size: 16-byte header + 16-byte payload = 32 bytes
    assert!(std::mem::size_of::<TaskEventMessage>() == 32);
};

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_message_sizes() {
        println!("MessageHeader size: {}", std::mem::size_of::<MessageHeader>());
        println!("DataPayload size: {}", std::mem::size_of::<DataPayload>());
        println!("TaskEventMessage size: {}", std::mem::size_of::<TaskEventMessage>());
        
        // Verify alignment
        assert!(std::mem::align_of::<TaskEventMessage>() <= 8);
    }

    #[test]
    fn test_message_creation() {
        let msg = TaskEventMessage::bool_data(123, true);
        assert!(msg.is_valid());
        assert_eq!(msg.header.task_id, 123);
        assert_eq!(msg.header.status, TaskStatus::SuccessWithData);
        assert_eq!(msg.header.data_type, DataType::Bool);
        assert!(msg.is_success());
        
        unsafe {
            assert_eq!(msg.payload.bool_val, true);
        }
    }

    #[test]
    fn test_serialization() {
        let msg = TaskEventMessage::u64_data(456, 0xDEADBEEF12345678);
        let serialized = MessageSerializer::serialize(&msg);
        
        // Verify serialized size is fixed at 32 bytes
        assert_eq!(serialized.len(), std::mem::size_of::<TaskEventMessage>());
        assert_eq!(serialized.len(), 32);
        
        unsafe {
            let deserialized = MessageSerializer::deserialize(
                serialized.as_ptr(),
                serialized.len(),
            ).unwrap();
            
            assert!(deserialized.is_valid());
            assert_eq!(deserialized.header.task_id, 456);
            assert_eq!(deserialized.header.data_type, DataType::U64);
            assert_eq!(deserialized.payload.u64_val, 0xDEADBEEF12345678);
        }
    }

    #[test]
    fn test_string_data() {
        let test_text = "Hello, Dart FFI Protocol!";
        let msg = TaskEventMessage::string_data(789, TaskStatus::SuccessWithData, test_text);
        
        assert!(msg.is_valid());
        assert_eq!(msg.header.task_id, 789);
        assert_eq!(msg.header.status, TaskStatus::SuccessWithData);
        assert_eq!(msg.header.data_type, DataType::String);
        assert!(msg.is_success());
        
        unsafe {
            assert_eq!(msg.payload.string.len, test_text.len());
            let string_slice = std::slice::from_raw_parts(msg.payload.string.ptr, msg.payload.string.len);
            assert_eq!(string_slice, test_text.as_bytes());
        }
    }

    #[test]
    fn test_error_message_convenience() {
        let error_msg = "Something went wrong";
        let msg = TaskEventMessage::error_message(999, TaskStatus::UnknownError, error_msg);
        
        assert!(msg.is_valid());
        assert!(msg.is_error());
        assert_eq!(msg.header.task_id, 999);
        assert_eq!(msg.header.status, TaskStatus::UnknownError);
        assert_eq!(msg.header.data_type, DataType::String);
    }

    #[test]
    fn test_zero_copy_design() {
        let test_data = vec![1, 2, 3, 4, 5];
        let data_ptr = test_data.as_ptr() as *mut u8;
        let data_len = test_data.len();
        
        // Create byte data message, only pass pointer
        let msg = TaskEventMessage::bytes_data(123, data_ptr, data_len);
        
        // Serialization should only be 32 bytes, not including actual data
        let serialized = MessageSerializer::serialize(&msg);
        assert_eq!(serialized.len(), 32);
        
        // Verify pointer and length are saved correctly
        unsafe {
            assert_eq!(msg.payload.bytes.ptr, data_ptr);
            assert_eq!(msg.payload.bytes.len, data_len);
            
            // Access original data through pointer
            if let Some((ptr, len)) = MessageSerializer::get_data_pointer(&msg) {
                let data_slice = std::slice::from_raw_parts(ptr, len);
                assert_eq!(data_slice, &test_data);
            } else {
                panic!("Should be able to get data pointer");
            }
        }
        
        // Test string zero-copy
        let test_string = "Zero copy string test";
        let string_msg = TaskEventMessage::string_data(456, TaskStatus::SuccessWithData, test_string);
        let string_serialized = MessageSerializer::serialize(&string_msg);
        
        // Also only 32 bytes
        assert_eq!(string_serialized.len(), 32);
        
        unsafe {
            if let Some((ptr, len)) = MessageSerializer::get_data_pointer(&string_msg) {
                let str_slice = std::slice::from_raw_parts(ptr, len);
                let recovered_string = std::str::from_utf8(str_slice).unwrap();
                assert_eq!(recovered_string, test_string);
            } else {
                panic!("Should be able to get string pointer");
            }
        }
    }
}

File: dart-quic/lib/src/protocol/binary_protocol.dart
/// High-efficiency binary protocol for Rust-Dart FFI communication
///
/// This is the Dart counterpart of the Rust binary_protocol.rs
/// Provides zero-copy message handling and efficient serialization/deserialization
library;

import 'dart:ffi';
import 'dart:typed_data';
import 'dart:convert';

/// Protocol version - for backward compatibility
const int protocolVersion = 1;

/// Protocol magic number
const int protocolMagic = 0xDABCFE01;

/// Task status enumeration
/// Corresponds to the Rust TaskStatus enum
enum TaskStatus {
  // === 0x0000 - 0x00FF ===
  success(0x0000), // Task completed successfully, no return data
  successWithData(0x0001), // Task completed successfully, has return data

  // === 0x0100 - 0x01FF ===
  workerShutdown(0x0100), // Worker thread shutdown normally

  // === 0x9000 - 0x9FFF ===
  unknownError(0x9001), // Unknown error

  // === 0xF000 - 0xFFFF ===
  protocolError(0xF001), // Protocol error
  versionMismatch(0xF002), // Version mismatch
  corruptedData(0xF003); // Data corrupted

  const TaskStatus(this.value);

  final int value;

  /// Create TaskStatus from integer value using factory constructor
  factory TaskStatus.fromValue(int value) {
    switch (value) {
      case 0x0000:
        return TaskStatus.success;
      case 0x0001:
        return TaskStatus.successWithData;
      case 0x0100:
        return TaskStatus.workerShutdown;
      case 0x9001:
        return TaskStatus.unknownError;
      case 0xF001:
        return TaskStatus.protocolError;
      case 0xF002:
        return TaskStatus.versionMismatch;
      case 0xF003:
        return TaskStatus.corruptedData;
      default:
        throw ArgumentError('Unknown TaskStatus value: $value');
    }
  }

  /// Check if this is a success status
  bool get isSuccess =>
      this == TaskStatus.success || this == TaskStatus.successWithData;

  /// Check if this is an error status
  bool get isError => !isSuccess && this != TaskStatus.workerShutdown;
}

/// Data type enumeration
/// Corresponds to the Rust DataType enum
enum DataType {
  none(0), // No data
  bool(1), // Boolean value
  u64(2), // 64-bit unsigned integer
  bytes(3), // Byte array (zero-copy pointer)
  string(4); // String (error messages, etc.)

  const DataType(this.value);

  final int value;

  /// Create DataType from integer value using factory constructor
  factory DataType.fromValue(int value) {
    switch (value) {
      case 0:
        return DataType.none;
      case 1:
        return DataType.bool;
      case 2:
        return DataType.u64;
      case 3:
        return DataType.bytes;
      case 4:
        return DataType.string;
      default:
        throw ArgumentError('Unknown DataType value: $value');
    }
  }
}

/// Message header structure (16 bytes)
/// Corresponds to the Rust MessageHeader struct
class MessageHeader {
  final int magic;
  final int version;
  final DataType dataType;
  final TaskStatus status;
  final int taskId;

  const MessageHeader({
    required this.magic,
    required this.version,
    required this.dataType,
    required this.status,
    required this.taskId,
  });

  /// Check if header is valid
  bool get isValid => magic == protocolMagic && version == protocolVersion;

  /// Serialize header to bytes (16 bytes)
  Uint8List toBytes() {
    final buffer = ByteData(16);
    int offset = 0;

    buffer.setUint32(offset, magic, Endian.little);
    offset += 4;
    buffer.setUint8(offset, version);
    offset += 1;
    buffer.setUint8(offset, dataType.value);
    offset += 1;
    buffer.setUint16(offset, status.value, Endian.little);
    offset += 2;
    buffer.setUint64(offset, taskId, Endian.little);
    offset += 8;

    return buffer.buffer.asUint8List();
  }

  /// Deserialize header from bytes
  static MessageHeader? fromBytes(Uint8List bytes) {
    if (bytes.length < 16) return null;

    final buffer = ByteData.sublistView(bytes);
    int offset = 0;

    final magic = buffer.getUint32(offset, Endian.little);
    offset += 4;
    final version = buffer.getUint8(offset);
    offset += 1;
    final dataTypeValue = buffer.getUint8(offset);
    offset += 1;
    final statusValue = buffer.getUint16(offset, Endian.little);
    offset += 2;
    final taskId = buffer.getUint64(offset, Endian.little);
    offset += 8;

    try {
      final dataType = DataType.fromValue(dataTypeValue);
      final status = TaskStatus.fromValue(statusValue);

      return MessageHeader(
        magic: magic,
        version: version,
        dataType: dataType,
        status: status,
        taskId: taskId,
      );
    } catch (e) {
      // Invalid enum values
      return null;
    }
  }

  @override
  String toString() {
    return 'MessageHeader(magic: 0x${magic.toRadixString(16)}, '
        'version: $version, dataType: $dataType, '
        'status: $status, taskId: $taskId)';
  }
}

/// Data payload for different types
/// Corresponds to the Rust DataPayload union
abstract class DataPayload {
  const DataPayload();

  /// Serialize payload to bytes (16 bytes)
  Uint8List toBytes();
}

/// No data payload
class NoDataPayload extends DataPayload {
  const NoDataPayload();

  @override
  Uint8List toBytes() {
    return Uint8List(16); // All zeros
  }
}

/// Boolean data payload
class BoolDataPayload extends DataPayload {
  final bool value;

  const BoolDataPayload(this.value);

  @override
  Uint8List toBytes() {
    final buffer = ByteData(16);
    buffer.setUint8(0, value ? 1 : 0);
    return buffer.buffer.asUint8List();
  }
}

/// U64 data payload
class U64DataPayload extends DataPayload {
  final int value;

  const U64DataPayload(this.value);

  @override
  Uint8List toBytes() {
    final buffer = ByteData(16);
    buffer.setUint64(0, value, Endian.little);
    return buffer.buffer.asUint8List();
  }
}

/// Bytes data payload (pointer + length)
class BytesDataPayload extends DataPayload {
  final Pointer<Uint8> ptr;
  final int length;

  const BytesDataPayload(this.ptr, this.length);

  @override
  Uint8List toBytes() {
    final buffer = ByteData(16);
    buffer.setUint64(0, ptr.address, Endian.little); // Pointer as address
    buffer.setUint64(8, length, Endian.little); // Length
    return buffer.buffer.asUint8List();
  }

  /// Get the actual byte data (zero-copy if possible)
  Uint8List getData() {
    if (ptr == nullptr || length == 0) {
      return Uint8List(0);
    }
    return ptr.asTypedList(length);
  }
}

/// String data payload (pointer + length)
class StringDataPayload extends DataPayload {
  final Pointer<Uint8> ptr;
  final int length;

  const StringDataPayload(this.ptr, this.length);

  @override
  Uint8List toBytes() {
    final buffer = ByteData(16);
    buffer.setUint64(0, ptr.address, Endian.little); // Pointer as address
    buffer.setUint64(8, length, Endian.little); // Length
    return buffer.buffer.asUint8List();
  }

  /// Get the actual string data
  String getString() {
    if (ptr == nullptr || length == 0) {
      return '';
    }
    final bytes = ptr.asTypedList(length);
    return utf8.decode(bytes);
  }
}

/// Complete task event message structure
/// Corresponds to the Rust TaskEventMessage struct (32 bytes total)
class QuicTaskMessage {
  final MessageHeader header;
  final DataPayload payload;

  const QuicTaskMessage({required this.header, required this.payload});

  /// Create no-data success message
  factory QuicTaskMessage.noData(int taskId) {
    return QuicTaskMessage(
      header: MessageHeader(
        magic: protocolMagic,
        version: protocolVersion,
        dataType: DataType.none,
        status: TaskStatus.success,
        taskId: taskId,
      ),
      payload: const NoDataPayload(),
    );
  }

  /// Create boolean data message
  factory QuicTaskMessage.boolData(int taskId, bool value) {
    return QuicTaskMessage(
      header: MessageHeader(
        magic: protocolMagic,
        version: protocolVersion,
        dataType: DataType.bool,
        status: TaskStatus.successWithData,
        taskId: taskId,
      ),
      payload: BoolDataPayload(value),
    );
  }

  /// Create U64 data message
  factory QuicTaskMessage.u64Data(int taskId, int value) {
    return QuicTaskMessage(
      header: MessageHeader(
        magic: protocolMagic,
        version: protocolVersion,
        dataType: DataType.u64,
        status: TaskStatus.successWithData,
        taskId: taskId,
      ),
      payload: U64DataPayload(value),
    );
  }

  /// Create bytes data message
  factory QuicTaskMessage.bytesData(
    int taskId,
    Pointer<Uint8> ptr,
    int length,
  ) {
    return QuicTaskMessage(
      header: MessageHeader(
        magic: protocolMagic,
        version: protocolVersion,
        dataType: DataType.bytes,
        status: TaskStatus.successWithData,
        taskId: taskId,
      ),
      payload: BytesDataPayload(ptr, length),
    );
  }

  /// Create string data message
  factory QuicTaskMessage.stringData(
    int taskId,
    TaskStatus status,
    String text,
  ) {
    // For simplicity, we'll use a different approach for string messages
    // In practice, you might want to allocate memory and create a pointer
    final bytes = utf8.encode(text);
    // This is a simplified version - in real implementation you'd need proper pointer management
    return QuicTaskMessage(
      header: MessageHeader(
        magic: protocolMagic,
        version: protocolVersion,
        dataType: DataType.string,
        status: status,
        taskId: taskId,
      ),
      payload: StringDataPayload(nullptr, bytes.length), // Simplified
    );
  }

  /// Create error message
  factory QuicTaskMessage.errorMessage(
    int taskId,
    TaskStatus errorType,
    String errorMsg,
  ) {
    return QuicTaskMessage.stringData(taskId, errorType, errorMsg);
  }

  /// Create worker shutdown message
  factory QuicTaskMessage.shutdownMessage() {
    return QuicTaskMessage(
      header: MessageHeader(
        magic: protocolMagic,
        version: protocolVersion,
        dataType: DataType.none,
        status: TaskStatus.workerShutdown,
        taskId: 0,
      ),
      payload: const NoDataPayload(),
    );
  }

  /// Check if the message is valid
  bool get isValid => header.isValid;

  /// Check if the message indicates success
  bool get isSuccess => header.status.isSuccess;

  /// Check if the message indicates an error
  bool get isError => header.status.isError;

  /// Get total message size (always 32 bytes)
  int get totalSize => 32;

  /// Serialize message to bytes (32 bytes)
  Uint8List toBytes() {
    final headerBytes = header.toBytes();
    final payloadBytes = payload.toBytes();

    final result = Uint8List(32);
    result.setRange(0, 16, headerBytes);
    result.setRange(16, 32, payloadBytes);

    return result;
  }

  /// Deserialize message from bytes
  static QuicTaskMessage? fromBytes(Uint8List bytes) {
    if (bytes.length < 32) return null;

    // Parse header
    final header = MessageHeader.fromBytes(bytes.sublist(0, 16));
    if (header == null || !header.isValid) return null;

    // Parse payload based on data type
    final payloadBytes = bytes.sublist(16, 32);
    final payloadBuffer = ByteData.sublistView(payloadBytes);

    DataPayload payload;

    switch (header.dataType) {
      case DataType.none:
        payload = const NoDataPayload();
        break;

      case DataType.bool:
        final boolValue = payloadBuffer.getUint8(0) != 0;
        payload = BoolDataPayload(boolValue);
        break;

      case DataType.u64:
        final u64Value = payloadBuffer.getUint64(0, Endian.little);
        payload = U64DataPayload(u64Value);
        break;

      case DataType.bytes:
        final ptrAddress = payloadBuffer.getUint64(0, Endian.little);
        final length = payloadBuffer.getUint64(8, Endian.little);
        final ptr = Pointer<Uint8>.fromAddress(ptrAddress);
        payload = BytesDataPayload(ptr, length);
        break;

      case DataType.string:
        final ptrAddress = payloadBuffer.getUint64(0, Endian.little);
        final length = payloadBuffer.getUint64(8, Endian.little);
        final ptr = Pointer<Uint8>.fromAddress(ptrAddress);
        payload = StringDataPayload(ptr, length);
        break;
    }

    return QuicTaskMessage(header: header, payload: payload);
  }

  @override
  String toString() {
    return 'QuicTaskMessage(header: $header, payload: ${payload.runtimeType})';
  }
}

/// High-performance message serializer
/// Corresponds to the Rust MessageSerializer
class MessageSerializer {
  const MessageSerializer._();

  /// Serialize message to binary data (fixed 32 bytes)
  static Uint8List serialize(QuicTaskMessage message) {
    return message.toBytes();
  }

  /// Deserialize binary data to message
  static QuicTaskMessage? deserialize(Uint8List data) {
    return QuicTaskMessage.fromBytes(data);
  }

  /// Get data pointed to by pointer (zero-copy access)
  /// Returns null if no pointer data is available
  static Uint8List? getDataPointer(QuicTaskMessage message) {
    switch (message.payload) {
      case BytesDataPayload(:final ptr, :final length):
        if (ptr != nullptr && length > 0) {
          return ptr.asTypedList(length);
        }
        break;
      case StringDataPayload(:final ptr, :final length):
        if (ptr != nullptr && length > 0) {
          return ptr.asTypedList(length);
        }
        break;
      default:
        break;
    }
    return null;
  }

  /// Get string data from pointer
  static String? getStringData(QuicTaskMessage message) {
    if (message.payload is StringDataPayload) {
      final stringPayload = message.payload as StringDataPayload;
      return stringPayload.getString();
    }
    return null;
  }
}

/// Protocol exception for error handling
class ProtocolException implements Exception {
  final String message;
  final TaskStatus? status;

  const ProtocolException(this.message, [this.status]);

  @override
  String toString() =>
      'ProtocolException: $message${status != null ? ' (status: $status)' : ''}';
}

/// Protocol utilities
class ProtocolUtils {
  const ProtocolUtils._();

  /// Validate protocol version compatibility
  static bool isVersionCompatible(int version) {
    return version == protocolVersion;
  }

  /// Create protocol error message
  static QuicTaskMessage createProtocolError(int taskId, String error) {
    return QuicTaskMessage.errorMessage(
      taskId,
      TaskStatus.protocolError,
      error,
    );
  }

  /// Create version mismatch error
  static QuicTaskMessage createVersionMismatchError(
    int taskId,
    int receivedVersion,
  ) {
    return QuicTaskMessage.errorMessage(
      taskId,
      TaskStatus.versionMismatch,
      'Version mismatch: expected $protocolVersion, got $receivedVersion',
    );
  }

  /// Create corrupted data error
  static QuicTaskMessage createCorruptedDataError(int taskId) {
    return QuicTaskMessage.errorMessage(
      taskId,
      TaskStatus.corruptedData,
      'Data corrupted',
    );
  }
}
